src/nbl/ui/CWindowXCB.cpp

```diff
@@ -0,0 +1,296 @@
+                    eventCallback->onWindowClosed(&m_window);
+                }
+                break;
+            }
+            case XCB_CLIENT_MESSAGE: {
+                xcb_client_message_event_t* cme = reinterpret_cast<xcb_client_message_event_t*>(event);
+                if(cme->data.data32[0] == MW_DELETE_WINDOW) {
+                    xcb.pxcb_unmap_window(m_window.getXcbConnection(), m_window.m_xcbWindow);
+                    xcb.pxcb_destroy_window(m_window.getXcbConnection(), m_window.m_xcbWindow);
+                    xcb.pxcb_flush(m_window.getXcbConnection());
+                    m_window.m_xcbWindow = 0;
+                    m_quit = true; // we need to quit the dispatch thread
+                    eventCallback->onWindowClosed(&m_window);
+                } else if(cme->data.data32[0] == NET_WM_PING && cme->window != connection->primaryScreen()->root) {
+                    xcb_client_message_event_t ev = *cme;
+                    ev.response_type = XCB_CLIENT_MESSAGE;
+                    ev.window = m_window.m_xcbWindow;
+                    ev.type = NET_WM_PING;
+                    xcb.pxcb_send_event(m_window.getXcbConnection(), 0, m_window.m_xcbWindow, XCB_EVENT_MASK_NO_EVENT, reinterpret_cast<const char*>(&ev));
+                    xcb.pxcb_flush(m_window.getXcbConnection());
+                }
+                break;
+            }
+        }
+        free(event);
+    }
+}
+
+
+void CWindowXCB::CDispatchThread::init()
+{
+
+}
+
+void CWindowXCB::CDispatchThread::exit()
+{
+}
+
+CWindowManagerXCB::CWindowManagerXCB() {
+}
+
+core::smart_refctd_ptr<IWindow> CWindowManagerXCB::createWindow(IWindow::SCreationParams&& creationParams)
+{
+    std::string title = std::string(creationParams.windowCaption);
+    auto window = core::make_smart_refctd_ptr<CWindowXCB>(core::smart_refctd_ptr<CWindowManagerXCB>(this), std::move(creationParams));
+    window->setCaption(title);
+    return window;
+}
+
+bool CWindowManagerXCB::setWindowSize_impl(IWindow* window, uint32_t width, uint32_t height) {
+    auto wnd = static_cast<IWindowXCB*>(window);
+    wnd->setWindowSize_impl(width, height);
+    return true;
+}
+
+bool CWindowManagerXCB::setWindowPosition_impl(IWindow* window, int32_t x, int32_t y) {
+    auto wnd = static_cast<IWindowXCB*>(window);
+    wnd->setWindowPosition_impl(x, y);
+    return true;
+}
+
+bool CWindowManagerXCB::setWindowRotation_impl(IWindow* window, bool landscape) {
+    auto wnd = static_cast<IWindowXCB*>(window);
+    wnd->setWindowRotation_impl(landscape);
+    return true;
+}
+
+bool CWindowManagerXCB::setWindowVisible_impl(IWindow* window, bool visible) {
+    auto wnd = static_cast<IWindowXCB*>(window);
+    wnd->setWindowVisible_impl(visible);
+    return true;
+}
+
+bool CWindowManagerXCB::setWindowMaximized_impl(IWindow* window, bool maximized) {
+    auto wnd = static_cast<IWindowXCB*>(window);
+    wnd->setWindowMaximized_impl(maximized);
+    return true;
+}
```

Comment:
why isn't this in its own `.cpp` file?

Reply:
I think I was copying the windowmanager, not sure why I followed along.